src/Controller/ResetPasswordController.php

```diff
@@ -71,63 +72,66 @@ public function checkEmail(SessionInterface $session): Response
      */
     public function reset(Request $request, UserPasswordEncoderInterface $passwordEncoder, string $token = null): Response
     {
-        //Put token in session and redirect to self
         if ($token) {
-            // We store token in session and remove it from the URL,
-            // to avoid any leak if someone get to know the URL (AJAX requests, Analytics...).
+            // We store the token in session and remove it from the URL, to avoid the URL being
+            // loaded in a browser and potentially leaking the token to 3rd party JavaScript.
             $this->storeTokenInSession($token);
 
             return $this->redirectToRoute('app_reset_password');
         }
 
-        //Get token out of session storage
         $token = $this->getTokenFromSession();
         if (null === $token) {
             throw $this->createNotFoundException();
         }
 
-        //Validate token using password helper
-        $user = $this->resetPasswordHelper->validateTokenAndFetchUser($token);
+        try {
+            $user = $this->resetPasswordHelper->validateTokenAndFetchUser($token);
+        } catch (ResetPasswordExceptionInterface $e) {
+            $this->addFlash('reset_password_error', \sprintf(
+                'There was a problem validating your reset request - %s',
+                $e->getReason()
+            ));
 
-        //Reset password after token verified
-        $form = $this->createForm(PasswordResettingFormType::class);
+            return $this->redirectToRoute('app_forgot_password_request');
```

Comment:
hmm, if validation fails we are redirecting to start a new request. May not be necessary, I.e. hiccup getting user from persistence..  perhaps redirect to reset with provided token. (security risk?)

either way, returned to reset a request from scratch wouldn't hurt. Not the ideal UX, but garbage collection would take care of any repercussions on the app side of things..

Reply:
> May not be necessary, I.e. hiccup getting user from persistence

Hmm. If there is a problem getting the User from persistence, then I can only think that it's because the User with that id doesn't exist anymore? I can't think of (but tell me if you disagree) a "normal" situation where something goes wrong here other than (more or less) a reset token was expired. And for that, we should redirect back to the start.